### ISSUE TITLE
Metadata ETL: Fill missing Override Cycles 

### ISSUE THREAD

**Issue by victorskl** [umccr/data-portal-apis#231]
Context:

- We have [_scheduled_ process](https://github.com/umccr/data-portal-apis/blob/dev/serverless.yml#L102) in-place to extract and load Lab Metadata into Portal database
- Lab Metadata tracking sheet is source of truth 
- We do not want to change the Lab tracking sheet on legacy metadata such as 2020 sheet
- However, we need to transform some data columns to conform with Portal workflow automation pipeline
- So, we need to correct these in Portal Metadata database
- This data transformation is typically deterministic steps
- We can arrange this by on-demands lambda invocation job and; i.e., also independent of _scheduled_ metadata process

Initial Case: **Fill missing Override Cycles** 
- Some legacy Lab Metadata rows contain non-filled Override Cycles
- Lookup by _minted_ "**Library ID**", they should all have matching Override Cycles (To be confirmed)
- As discussed in #229; it is an Error condition in Portal to drive Workflow Automation if Override Cycles is missing for a sample metadata lookup by _minted_ "**Library ID**"

Analysis:
- Lookup by `L2000200` in Lab tracking sheet
- Lookup in Portal DB
```
select * from data_portal.data_portal_labmetadata where library_id like 'L2000200';
select * from data_portal.data_portal_labmetadata where library_id like '%L2000200%';

-- number of minted_library_id occurrence frequency
select substring(library_id, 1, 8) as minted_library_id, count(1) from data_portal.data_portal_labmetadata group by minted_library_id having count(1) > 1;

select * from data_portal.data_portal_labmetadata where library_id like '%L2000116%';
```

Plan (Draft propose plan):
- Transform only when 
  - Override Cycles is null; AND
  - at least one record in lookup by `minted_library_id` aggregate is having Override Cycles; AND
  - is matched by list of `minted_library_id` provided by user input

NOTE: Transformation use cases / requirement may grow.

**Comment by victorskl** [comment]
We can start with SEQ-2 (validation) Run Samples as pilot case.

**Comment by ohofmann** [comment]
I general I'd say it is okay to throw an error and stop converting (or aligning, etc.) if any of the key metadata is missing. I'd rather go into the habit of correcting the tracking sheet or Clarity in the future than trying to impute missing values.

**Comment by reisingerf** [comment]
Agreed. The fix should be in the source not the data portal.
Are there any cases we cannot (currently) fix in the tracking sheet?

**Comment by victorskl** [comment]
Sure, sound good. Fixing at source is the best case scenario, yes. 

We can probably start with SEQ-2 (validation) Run Samples as follows. All the minted Library ID (e.g. `L2000248`, `L2000200`, etc) need to populate Override Cycles, if there are missing one.

Then, Portal can continue perform SEQ-2 validation run on code changes and testing the pipeline. Otherwise this block validation run.

```
{
  'PTC_TsqN200327_L2000248_topup',
  'PTC_TsqN200511_L2000431_topup',
  'PTC_TSO200609VD_L2000552',
  'PTC_TsqN2000504_L2000416_topup',
  'PTC_SsCRE200330KC_L2000185_topup',
  'PTC_SsCRE200323VD_L2000199_topup',
  'PTC_SsCRE200323VD_L2000200_topup',
  'PTC_SsCRE200323VD_L2000198_topup',
  'PTC_TsqN200327_L2000243_topup',
  'PTC_SsCRE200323LL_L2000173_topup',
  'PTC_TsqN200511_L2000436',
  'PTC_TsqN2000504_L2000411_topup',
  'PTC_TSO200609VD_L2000553',
  'PTC_TSO200609VD_L2000555',
  'NTC_TSO200609VD_L2000559',
  'PTC_TsqN200511_L2000433',
  'PTC_SsCRE200330VD_L2000212_topup',
  'PTC_TsqN2000504_L2000412_topup',
  'NTC_TsqN200511_L2000438',
  'PTC_SsCRE200323LL_L2000172_topup',
  'PTC_TSO200609VD_L2000554',
  'PTC_TsqN200511_L2000432',
  'PTC_SsCRE200330KC_L2000187_topup',
  'PTC_SsCRE200323LL_L2000176_topup',
  'PTC_TsqN2000504_L2000413_topup',
  'PTC_TsqN2000504_L2000414_topup',
  'PTC_TSO200609VD_L2000558',
  'PTC_TSO200609VD_L2000556',
  'PTC_TsqN2000504_L2000415_topup',
  'PTC_SsCRE200330KC_L2000186_topup',
  'PTC_TSO200609VD_L2000557',
  'PTC_TsqN200511_L2000434',
  'PTC_SsCRE200323LL_L2000176_topup2',
  'PTC_SsCRE200323LL_L2000177_topup',
  'PTC_TsqN200511_L2000435',
  'PTC_SsCRE200323LL_L2000174_topup',
  'PTC_SsCRE200323LL_L2000175_topup'
}
```

**Comment by reisingerf** [comment]
OK. I can copy the missing `OverrideCycles` from the `topup` rows to the "minted" ones.

**Comment by ohofmann** [comment]
Can I have a quick look first? The overridecycles might differ between topup and original library.

**Comment by reisingerf** [comment]
Ah, already half way through, but can revert if needed.
I also noticed some differences. E.g. L2000433 with has a different `OverrrideCycle` for `L2000433_rerun`. Not sure if that is OK or needs adjusting.
I'll pause the update for now....

**Comment by ohofmann** [comment]
Depending on which run they were on they'd have different overrides, yes - we may need the lab to visit this, or go through each run's RunInfo.XML to check the settings and generate the cycles based on that. Do we need to include the 2019/2020 SEQC-II controls, or would it be okay to limit this to samples from the start of 2021?

**Comment by victorskl** [comment]
> Do we need to include the 2019/2020 SEQC-II controls, or would it be okay to limit this to samples from the start of 2021?

Ah yes; that's another way to solve it, too. It does not need to be 2019/2020 SEQC-II controls dataset. It is okay to limit this to samples from the start of 2021. Smaller sample size is even better; so that faster testing turn around.

Working out with @reisingerf on next appropriate run data to use for this validation test run, meanwhile.

**Comment by victorskl** [comment]
Fixed by ICA link data + PR #241 to fallback graceful handle legacy data.
